src/Orleans/Async/TaskExtensions.cs

```diff
@@ -130,11 +131,13 @@ internal static T WaitForResultWithThrow<T>(this Task<T> task, TimeSpan timeout)
                 return;
             }
 
-            await Task.WhenAny(taskToComplete, Task.Delay(timeout));
+            var timeoutCancellationTokenSource = new CancellationTokenSource();
+            var completedTask = await Task.WhenAny(taskToComplete, Task.Delay(timeout, timeoutCancellationTokenSource.Token));
 
             // We got done before the timeout, or were able to complete before this code ran, return the result
-            if (taskToComplete.IsCompleted)
+            if (taskToComplete == completedTask)
```

Comment:
No, because  <code>Task.WhenAny</code> returns <code>Task\<Task\></code> where the child Task represents whichever Task completed first, and awaiting it only unwraps the child task, thus the state of the task that possibly ended in the canceled or faulted state isn't being observed.
For exception of the <code>taskToComplete</code> to be thrown there is the following line: 
<code>return await taskToComplete;</code>